File: system/ui/installer.py
#!/usr/bin/env python3
import io
import os
import subprocess
import sys
import threading
import time
from dataclasses import dataclass
from enum import IntEnum
from importlib.resources import files
import pyray as rl

from cereal import log
from openpilot.common.swaglog import cloudlog
from openpilot.common.time_helpers import system_time_valid
from openpilot.system.hardware import HARDWARE
from openpilot.system.ui.lib.application import gui_app, FontWeight
from openpilot.system.ui.lib.button import gui_button, ButtonStyle
from openpilot.system.ui.lib.label import gui_label, gui_text_box

NETWORK_TYPE = log.DeviceState.NetworkType

# TODO: can we patch these strings
GIT_URL = "https://github.com/commaai/openpilot.git"
BRANCH = "master"

GIT_SSH_URL = "git@github.com:commaai/openpilot.git"
CONTINUE_PATH = "/data/continue.sh"

CACHE_PATH = "/data/openpilot.cache"

INSTALL_PATH = "/data/openpilot"
TMP_INSTALL_PATH = "/data/tmppilot"


class InstallStage(IntEnum):
  CLONING = 0
  RECEIVING_OBJECTS = 1
  RESOLVING_DELTAS = 2
  UPDATING_FILES = 3
  FINALIZING = 4
  COMPLETED = 5


# Stage weights for progress calculation
STAGE_WEIGHTS = [
  ("Receiving objects: ", 91),
  ("Resolving deltas: ", 2),
  ("Updating files: ", 7),
]


@dataclass
class StateInstalling:
  stage = InstallStage.CLONING
  progress = 0


@dataclass
class StateDone:
  pass


@dataclass
class StateError:
  error: str


class Installer:
  def __init__(self):
    self.state = StateInstalling()
    self.install_thread = threading.Thread(target=self.do_install, daemon=True)
    self.install_thread.start()

  def run_command(self, cmd):
    try:
      subprocess.run(cmd, check=True, shell=True)
      return True
    except subprocess.CalledProcessError as e:
      self.state = StateError(f"Command failed: {cmd}\nError: {e}")
      return False

  def run_git_command(self, args, cwd=None):
    proc = subprocess.Popen(["git"] + args, cwd=cwd, stderr=subprocess.PIPE, universal_newlines=True)
    self.read_progress(proc)
    ret = proc.wait()
    return ret == 0

  def read_progress(self, proc: subprocess.Popen):
    if proc.stderr is None:
      return
    base = 0
    for line in io.TextIOWrapper(proc.stderr, encoding="utf-8"):
      for prefix, weight in STAGE_WEIGHTS:
        if line.startswith(prefix):
          perc = line.split(prefix)[1].split("%")[0]
          p = base + int(float(perc) / 100.0 * weight)
          self.progress = p
          break
        base += weight

  def do_install(self):
    # wait for valid time
    while not system_time_valid():
      time.sleep(0.5)
      cloudlog.debug("Waiting for valid time")

    # cleanup previous install attempts
    if not self.run_command(f"rm -rf {TMP_INSTALL_PATH} {INSTALL_PATH}"):
      return

    # do the install
    if os.path.exists(CACHE_PATH):
      self.cached_fetch(CACHE_PATH)
    else:
      self.fresh_clone()

  def fresh_clone(self):
    cloudlog.debug("Doing fresh clone")
    success = self.run_git_command(["clone", "--progress", GIT_URL, "-b", BRANCH, "--depth=1", "--recurse-submodules", TMP_INSTALL_PATH])
    if not success:
      return

    self.finish_install()

  def cached_fetch(self, cache: str):
    cloudlog.debug(f"Fetching with cache: {cache}")
    if not self.run_command(f"cp -rp {cache} {TMP_INSTALL_PATH}"):
      return

    os.chdir(TMP_INSTALL_PATH)
    if not self.run_command(f"git remote set-branches --add origin {BRANCH}"):
      return

    self.progress = 10
    success = self.run_git_command(["fetch", "--progress", "origin", BRANCH], cwd=TMP_INSTALL_PATH)
    if not success:
      return

    self.finish_install()

  def finish_install(self):
    self.progress = 100
    self.stage = InstallStage.FINALIZING

    # Ensure correct branch
    os.chdir(TMP_INSTALL_PATH)
    if not self.run_command(f"git checkout {BRANCH}"):
      return
    if not self.run_command(f"git reset --hard origin/{BRANCH}"):
      return
    if not self.run_command("git submodule update --init"):
      return

    # move into place
    if not self.run_command(f"mv {TMP_INSTALL_PATH} {INSTALL_PATH}"):
      return

    # Set up SSH keys for internal builds if needed
    if os.environ.get('INTERNAL', '0') == '1':
      if not self.run_command("mkdir -p /data/params/d/"):
        return

      # https://github.com/commaci2.keys
      ssh_keys = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIMX2kU8eBZyEWmbq0tjMPxksWWVuIV/5l64GabcYbdpI"
      params = {
        "SshEnabled": "1",
        "RecordFrontLock": "1",
        "GithubSshKeys": ssh_keys,
      }

      for key, value in params.items():
        with open(f"/data/params/d/{key}", "w") as param:
          param.write(value)

      if not self.run_command(
        f"cd {INSTALL_PATH} && "
        f"git remote set-url origin --push {GIT_SSH_URL} && "
        f"git config --replace-all remote.origin.fetch \"+refs/heads/*:refs/remotes/origin/*\""
      ):
        return

    # Copy continue.sh script
    with files("openpilot.selfdrive.ui").joinpath("installer/continue_openpilot.sh").open("rb") as rf:
      with open("/data/continue.sh.new", "wb") as wf:
        wf.write(rf.read())

    if not self.run_command("chmod +x /data/continue.sh.new"):
      return
    if not self.run_command(f"mv /data/continue.sh.new {CONTINUE_PATH}"):
      return

    self.stage = InstallStage.COMPLETED
    self.state = StateDone()

    # Exit after 60 seconds
    threading.Timer(60, sys.exit).start()

  def render(self, rect: rl.Rectangle):
    match self.state:
      case StateInstalling():
        self.render_installing(rect)
      case StateDone():
        self.render_done(rect)
      case StateError(error):
        self.render_error(rect, error)

  def render_installing(self, rect: rl.Rectangle):
    title_rect = rl.Rectangle(rect.x + 150, rect.y + 290, rect.width - 300, 90)
    gui_label(title_rect, "Installing...", 90, font_weight=FontWeight.SEMI_BOLD)

    bar_rect = rl.Rectangle(rect.x + 150, rect.y + 290 + 170, rect.width - 300, 72)
    rl.draw_rectangle_rounded(bar_rect, 0.05, 10, rl.Color(41, 41, 41, 255))

    if self.progress > 0:
      chunk_width = (rect.width - 300) * self.progress / 100
      chunk_rect = rl.Rectangle(rect.x + 150, rect.y + 290 + 170, chunk_width, 72)
      rl.draw_rectangle_rounded(chunk_rect, 0.05, 10, rl.Color(54, 77, 239, 255))

    percent_rect = rl.Rectangle(rect.x + 150, rect.y + 290 + 170 + 102, rect.width - 300, 70)
    gui_label(percent_rect, f"{self.progress}%", 70, font_weight=FontWeight.LIGHT)

  def render_done(self, rect: rl.Rectangle):
    title_rect = rl.Rectangle(rect.x + 150, rect.y + 290, rect.width - 300, 90)
    gui_label(title_rect, "Installation Complete", 90, font_weight=FontWeight.SEMI_BOLD)

    message_rect = rl.Rectangle(rect.x + 150, rect.y + 290 + 170, rect.width - 300, 72)
    gui_label(message_rect, "openpilot will start soon...", 70, font_weight=FontWeight.LIGHT)

  def render_error(self, rect: rl.Rectangle, error: str):
    title_rect = rl.Rectangle(rect.x + 150, rect.y + 200, rect.width - 300, 90)
    gui_label(title_rect, "Installation Error", 90, rl.Color(255, 89, 79, 255), font_weight=FontWeight.SEMI_BOLD)

    error_rect = rl.Rectangle(rect.x + 150, rect.y + 200 + 120, rect.width - 300, rect.height - 400)
    gui_text_box(error_rect, error, 50)

    button_rect = rl.Rectangle(rect.x + 150, rect.height - 150, rect.width - 300, 100)
    if gui_button(button_rect, "Reboot Device", button_style=ButtonStyle.PRIMARY):
      HARDWARE.reboot()


def main():
  try:
    gui_app.init_window("Installer")
    installer = Installer()
    for _ in gui_app.render():
      installer.render(rl.Rectangle(0, 0, gui_app.width, gui_app.height))
  except Exception as e:
    print(f"Installer error: {e}")
  finally:
    gui_app.close()


if __name__ == "__main__":
  main()

File: openpilot/selfdrive/ui/installer/installer.cc
#include <array>
#include <atomic>
#include <cassert>
#include <fstream>
#include <map>

#include "common/swaglog.h"
#include "common/util.h"
#include "common/hardware/hw.h"
#include "json11/json11.hpp"
#include "raylib.h"

int freshClone();
int cachedFetch(const std::string &cache);
int executeGitCommand(const std::string &cmd);
int downloadLfsFiles();

std::string get_str(std::string const s) {
  std::string::size_type pos = s.find('?');
  assert(pos != std::string::npos);
  return s.substr(0, pos);
}

// Leave some extra space for the fork installer
const std::string GIT_URL = get_str("https://github.com/commaai/openpilot.git" "?                                                                ");
const std::string BRANCH_STR = get_str(BRANCH "?                                                                ");

#define GIT_SSH_URL "git@github.com:commaai/openpilot.git"
#define CONTINUE_PATH "/tmp/continue.sh"

const std::string INSTALL_PATH = "/tmp/openpilot";
const std::string VALID_CACHE_PATH = "/tmp/.openpilot_cache";
const std::string LFS_PROGRESS_PATH = "/tmp/openpilot_lfs_progress";

#define TMP_INSTALL_PATH "/tmp/tmppilot"

const int FONT_SIZE = 160;

#define BINARY_SYMBOL(path) "_binary_openpilot_" path

extern const uint8_t str_continue[] asm(BINARY_SYMBOL("selfdrive_ui_installer_continue_openpilot_sh_start"));
extern const uint8_t str_continue_end[] asm(BINARY_SYMBOL("selfdrive_ui_installer_continue_openpilot_sh_end"));
extern const uint8_t inter_ttf[] asm(BINARY_SYMBOL("selfdrive_ui_installer_inter_ascii_ttf_start"));
extern const uint8_t inter_ttf_end[] asm(BINARY_SYMBOL("selfdrive_ui_installer_inter_ascii_ttf_end"));
extern const uint8_t inter_light_ttf[] asm(BINARY_SYMBOL("selfdrive_assets_fonts_Inter_Light_ttf_start"));
extern const uint8_t inter_light_ttf_end[] asm(BINARY_SYMBOL("selfdrive_assets_fonts_Inter_Light_ttf_end"));
extern const uint8_t inter_bold_ttf[] asm(BINARY_SYMBOL("selfdrive_assets_fonts_Inter_Bold_ttf_start"));
extern const uint8_t inter_bold_ttf_end[] asm(BINARY_SYMBOL("selfdrive_assets_fonts_Inter_Bold_ttf_end"));

Font font_inter;
Font font_roman;
Font font_display;

const bool tici_device = Hardware::get_device_type() == cereal::InitData::DeviceType::TICI ||
                         Hardware::get_device_type() == cereal::InitData::DeviceType::TIZI;

std::vector<std::string> tici_prebuilt_branches = {"release3", "release-tici", "release3-staging", "nightly", "nightly-dev"};
std::string migrated_branch;

void branchMigration() {
  migrated_branch = BRANCH_STR;
  cereal::InitData::DeviceType device_type = Hardware::get_device_type();
  if (device_type == cereal::InitData::DeviceType::TICI) {
    if (std::find(tici_prebuilt_branches.begin(), tici_prebuilt_branches.end(), BRANCH_STR) != tici_prebuilt_branches.end()) {
      migrated_branch = "release-tici";
    } else if (BRANCH_STR == "master") {
      migrated_branch = "master-tici";
    }
  } else if (device_type == cereal::InitData::DeviceType::TIZI) {
    if (BRANCH_STR == "release3") {
      migrated_branch = "release-tizi";
    } else if (BRANCH_STR == "release3-staging") {
      migrated_branch = "release-tizi-staging";
    }
  } else if (device_type == cereal::InitData::DeviceType::MICI) {
    if (BRANCH_STR == "release3") {
      migrated_branch = "release-mici";
    } else if (BRANCH_STR == "release3-staging") {
      migrated_branch = "release-mici-staging";
    }
  }
}

void run(const char* cmd) {
  int err = std::system(cmd);
  assert(err == 0);
}

void finishInstall() {
  BeginDrawing();
    ClearBackground(BLACK);
    if (tici_device) {
      const char *m = "Finishing install...";
      int text_width = MeasureText(m, FONT_SIZE);
      DrawTextEx(font_display, m, (Vector2){(float)(GetScreenWidth() - text_width)/2 + FONT_SIZE, (float)(GetScreenHeight() - FONT_SIZE)/2}, FONT_SIZE, 0, WHITE);
    } else {
      DrawTextEx(font_display, "finishing setup", (Vector2){12, 0}, 77, 0, (Color){255, 255, 255, (unsigned char)(255 * 0.9)});
    }
  EndDrawing();
  util::sleep_for(60 * 1000);
}

void renderProgress(int progress) {
  BeginDrawing();
    ClearBackground(BLACK);
    if (tici_device) {
      DrawTextEx(font_inter, "Installing...", (Vector2){150, 290}, 110, 0, WHITE);
      Rectangle bar = {150, 570, (float)GetScreenWidth() - 300, 72};
      DrawRectangleRec(bar, (Color){41, 41, 41, 255});
      progress = std::clamp(progress, 0, 100);
      bar.width *= progress / 100.0f;
      DrawRectangleRec(bar, (Color){70, 91, 234, 255});
      DrawTextEx(font_inter, (std::to_string(progress) + "%").c_str(), (Vector2){150, 670}, 85, 0, WHITE);
    } else {
      DrawTextEx(font_display, "installing...", (Vector2){12, 0}, 77, 0, (Color){255, 255, 255, (unsigned char)(255 * 0.9)});
      const std::string percent_str = std::to_string(progress) + "%";
      DrawTextEx(font_inter, percent_str.c_str(), (Vector2){12, (float)(GetScreenHeight() - 154 + 20)}, 154, 0,
                 (Color){255, 255, 255, (unsigned char)(255 * 0.9 * 0.65)});
    }

  EndDrawing();
}

int doInstall() {
  // wait for valid time
  while (!util::system_time_valid()) {
    util::sleep_for(500);
    LOGD("Waiting for valid time");
  }

  // cleanup previous install attempts
  run("rm -rf " TMP_INSTALL_PATH);

  // do the install
  if (util::file_exists(INSTALL_PATH) && util::file_exists(VALID_CACHE_PATH)) {
    return cachedFetch(INSTALL_PATH);
  } else {
    return freshClone();
  }
}

int freshClone() {
  LOGD("Doing fresh clone");
  std::string cmd = util::string_format("GIT_LFS_SKIP_SMUDGE=1 git clone --progress %s -b %s --depth=1 %s 2>&1",
                                        GIT_URL.c_str(), migrated_branch.c_str(), TMP_INSTALL_PATH);
  return executeGitCommand(cmd);
}

int cachedFetch(const std::string &cache) {
  LOGD("Fetching with cache: %s", cache.c_str());

  run(util::string_format("cp -rp %s %s", cache.c_str(), TMP_INSTALL_PATH).c_str());
  run(util::string_format("cd %s && git remote set-url origin %s", TMP_INSTALL_PATH, GIT_URL.c_str()).c_str());
  run(util::string_format("cd %s && git remote set-branches --add origin %s", TMP_INSTALL_PATH, migrated_branch.c_str()).c_str());

  renderProgress(10);

  return executeGitCommand(util::string_format("cd %s && git fetch --progress origin %s 2>&1", TMP_INSTALL_PATH, migrated_branch.c_str()));
}

int executeGitCommand(const std::string &cmd) {
  static const std::array stages = {
    // prefix, weight in percentage
    std::pair{"Receiving objects: ", 1},
    std::pair{"Resolving deltas: ", 0},
    std::pair{"Updating files: ", 0},
  };

  FILE *pipe = popen(cmd.c_str(), "r");
  if (!pipe) {
    return -1;
  }

  int installer_progress = 0;
  auto process_line = [&](const std::string &line) {
    size_t latest_stage_pos = std::string::npos;
    int latest_stage_base = 0;
    int latest_stage_weight = 0;
    int base = 0;
    for (const auto &[text, weight] : stages) {
      size_t stage_pos = line.rfind(text);
      if (stage_pos != std::string::npos && (latest_stage_pos == std::string::npos || stage_pos > latest_stage_pos)) {
        latest_stage_pos = stage_pos;
        latest_stage_base = base;
        latest_stage_weight = weight;
      }
      base += weight;
    }

    if (latest_stage_pos != std::string::npos) {
      size_t percent_pos = line.find("%", latest_stage_pos);
      if (percent_pos != std::string::npos && percent_pos >= 3) {
        int percent = std::stoi(line.substr(percent_pos - 3, 3));
        installer_progress = latest_stage_base + int(percent / 100. * latest_stage_weight);
        renderProgress(installer_progress);
      }
    }

    printf("%s\n", line.c_str());
    printf("Installer progress: %d%%\n\n", installer_progress);
    fflush(stdout);
  };

  std::string line;
  int character;
  while ((character = fgetc(pipe)) != EOF) {
    if (character == '\r' || character == '\n') {
      if (!line.empty()) {
        process_line(line);
        line.clear();
      }
    } else {
      line += character;
    }
  }
  if (!line.empty()) {
    process_line(line);
  }
  return pclose(pipe);
}

int downloadLfsFiles() {
  std::string parse_error;
  json11::Json lfs_files = json11::Json::parse(util::check_output("git lfs ls-files --json"), parse_error);
  if (!parse_error.empty()) {
    LOGE("Failed to parse git lfs ls-files: %s", parse_error.c_str());
    return -1;
  }

  uint64_t total_bytes = 0;
  std::map<std::string, uint64_t> downloaded_bytes;
  for (const auto &file : lfs_files["files"].array_items()) {
    const std::string name = file["name"].string_value();
    const uint64_t size = file["size"].number_value();
    total_bytes += size;
    downloaded_bytes[name] = file["downloaded"].bool_value() ? size : 0;
  }

  int installer_progress = 1;
  auto process_line = [&](const std::string &line) {
    size_t direction_end = line.find(' ');
    size_t file_count_end = line.find(' ', direction_end + 1);
    size_t byte_count_end = line.find(' ', file_count_end + 1);
    if (direction_end != std::string::npos && file_count_end != std::string::npos && byte_count_end != std::string::npos &&
        line.substr(0, direction_end) == "download") {
      const std::string byte_count = line.substr(file_count_end + 1, byte_count_end - file_count_end - 1);
      size_t slash = byte_count.find('/');
      if (slash != std::string::npos) {
        const std::string name = line.substr(byte_count_end + 1);
        downloaded_bytes[name] = std::stoull(byte_count.substr(0, slash));

        uint64_t completed_bytes = 0;
        for (const auto &[path, bytes] : downloaded_bytes) {
          completed_bytes += bytes;
        }
        if (total_bytes > 0) {
          int next_progress = 1 + int(98. * completed_bytes / total_bytes);
          if (next_progress != installer_progress) {
            installer_progress = next_progress;
            renderProgress(installer_progress);
            printf("%s\n", line.c_str());
            printf("Installer progress: %d%%\n\n", installer_progress);
            fflush(stdout);
          }
        }
      }
    }
  };

  unlink(LFS_PROGRESS_PATH.c_str());
  std::atomic<bool> lfs_finished = false;
  int lfs_result = -1;
  std::thread lfs_thread([&]() {
    lfs_result = std::system(("GIT_LFS_PROGRESS=" + LFS_PROGRESS_PATH + " git lfs pull").c_str());
    lfs_finished = true;
  });

  std::ifstream progress_file;
  auto read_progress = [&]() {
    if (!progress_file.is_open()) {
      progress_file.clear();
      progress_file.open(LFS_PROGRESS_PATH);
    }
    if (progress_file.is_open()) {
      std::string line;
      while (std::getline(progress_file, line)) {
        process_line(line);
      }
      progress_file.clear();
    }
  };

  while (!lfs_finished) {
    read_progress();
    util::sleep_for(100);
  }
  lfs_thread.join();
  read_progress();
  unlink(LFS_PROGRESS_PATH.c_str());
  return lfs_result;
}

void cloneFinished(int exitCode) {
  LOGD("git finished with %d", exitCode);
  assert(exitCode == 0);

  // ensure correct branch is checked out
  int err = chdir(TMP_INSTALL_PATH);
  assert(err == 0);
  run(("GIT_LFS_SKIP_SMUDGE=1 git checkout " + migrated_branch).c_str());
  run(("GIT_LFS_SKIP_SMUDGE=1 git reset --hard origin/" + migrated_branch).c_str());
//   run("git submodule update --init");

  int lfs_result = downloadLfsFiles();
  assert(lfs_result == 0);
  renderProgress(100);

  // move into place
  run(("rm -f " + VALID_CACHE_PATH).c_str());
  run(("rm -rf " + INSTALL_PATH).c_str());
  run(util::string_format("mv %s %s", TMP_INSTALL_PATH, INSTALL_PATH.c_str()).c_str());

// #ifdef INTERNAL
//   run("mkdir -p /data/params/d/");
//
//   // https://github.com/commaci2.keys
//   const std::string ssh_keys = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIMX2kU8eBZyEWmbq0tjMPxksWWVuIV/5l64GabcYbdpI";
//   std::map<std::string, std::string> params = {
//     {"SshEnabled", "1"},
//     {"RecordFrontLock", "1"},
//     {"GithubSshKeys", ssh_keys},
//   };
//   for (const auto& [key, value] : params) {
//     std::ofstream param;
//     param.open("/data/params/d/" + key);
//     param << value;
//     param.close();
//   }
//   run(("cd " + INSTALL_PATH + " && "
//       "git remote set-url origin --push " GIT_SSH_URL " && "
//       "git config --replace-all remote.origin.fetch \"+refs/heads/*:refs/remotes/origin/*\"").c_str());
// #endif

  // write continue.sh
  FILE *of = fopen("/tmp/continue.sh.new", "wb");
  assert(of != NULL);

  size_t num = str_continue_end - str_continue;
  size_t num_written = fwrite(str_continue, 1, num, of);
  assert(num == num_written);
  fclose(of);

//   run("chmod +x /data/continue.sh.new");
//   run("mv /data/continue.sh.new " CONTINUE_PATH);

  // wait for the installed software's UI to take over
  finishInstall();
}

int main(int argc, char *argv[]) {
  if (tici_device) {
    InitWindow(2160, 1080, "Installer");
  } else {
    InitWindow(536, 240, "Installer");
  }

  font_inter = LoadFontFromMemory(".ttf", inter_ttf, inter_ttf_end - inter_ttf, FONT_SIZE, NULL, 0);
  font_roman = LoadFontFromMemory(".ttf", inter_light_ttf, inter_light_ttf_end - inter_light_ttf, FONT_SIZE, NULL, 0);
  font_display = LoadFontFromMemory(".ttf", inter_bold_ttf, inter_bold_ttf_end - inter_bold_ttf, FONT_SIZE, NULL, 0);
  SetTextureFilter(font_inter.texture, TEXTURE_FILTER_BILINEAR);
  SetTextureFilter(font_roman.texture, TEXTURE_FILTER_BILINEAR);
  SetTextureFilter(font_display.texture, TEXTURE_FILTER_BILINEAR);

  branchMigration();

  if (util::file_exists(CONTINUE_PATH)) {
    finishInstall();
  } else {
    renderProgress(0);
    int result = doInstall();
    cloneFinished(result);
  }

  CloseWindow();
  UnloadFont(font_inter);
  UnloadFont(font_roman);
  UnloadFont(font_display);
  return 0;
}
